### PR TITLE
Add eleventy to automatically build web pages

### LINKED ISSUE #10: Build system for publishing scores to a website
## Why
We want to be able to publish the songs to the web, making the scores and MIDI-files easily accessible.

## How
Using Github Actions, this build process will be run on merge or commit to master. This will happen in the following steps:

1. For run Lilypond for each song and output:
  - MIDI-files
  - SVG file with the score
  - JSON file containing title, SVG file name, MIDI file names
 2. Output these to a set directory structure
 3. Use 11ty to generate a set of static pages
bild

Things to consider:
- How do we extract the title? Suggested solution is to use `\bookOutputName` for the file name
- How do we get a searchable output?

## Todos
- [ ] Generate proper output from Lilypond in a Github Actions https://github.com/Snapsakademien/scores-ly/pull/9
- [ ] #12

### LINKED ISSUE #12: Use 11ty to generate a static website
Idea: Add json file to every song folder with metadata, and also name of output SVGs and MIDIs.

### PR DESCRIPTION
**Pull request by tskarhed**
- Adds basic templates
- Adds song frontmatter structure

Midi and svg output names must be written in the files.

nodejs and 11ty required.
[Screencast from 08-05-23 17:28:40.webm](https://user-images.githubusercontent.com/1438972/236865583-9638eb30-5eb1-42d1-bd8b-b0841fe97f46.webm)

How to run:

1. Install nodejs
2. Install the latest version of Lilypond
3. Run `npm install` in the repository
4. Build lilypond output by running `npm run build:songs`
5. Build 11ty files with `npm run build:pre-11ty`
6. Start server `npm start`
7. Visit `localhost:8080` and you have the website

Closes #10
Closes #12

### CODE REVIEW

File: songs/Angbatssang/index.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+svgFiles:
+    - Angbatssang-1.svg
+    - Angbatssang-2.svg
+    - Angbatssang-3.svg
+midiFiles:
+    - main.midi
+    - main-1.midi
+    - main-2.midi
+    - main-3.midi
+    - main-4.midi
```
Comment (tskarhed) [inline songs/Angbatssang/index.md:13]: Names should be unique, but I guess this works

File: songs/Angbatssang.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+title: Ångbåtssång
+author: Otto Lindblad
+midiFiles: []
+svgFiles:
+  - songs/Angbatssang-1.svg
+  - songs/Angbatssang-2.svg
+  - songs/Angbatssang-3.svg
+  - songs/AngbatssangEnd.svg
+---
```
Comment (tskarhed) [inline songs/Angbatssang.md:10]: Suggestion: ignore md files and see them as build artifacts. Instead, have just a yaml file with title and author, from which a script adds svgFiles and midiFiels

File: songs/styles.css [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+.song_list__title {
+    font-family: var(--heading-font);
+    font-size: 1.6rem;
+}
+
+.song_list__author {
+    overflow: hidden;
+    white-space: nowrap;
+    text-overflow: ellipsis;
+}
```
Comment (tskarhed) [inline songs/styles.css:68]:
```
\n
```
:tada:

File: .eleventy.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+const yaml = require("js-yaml");
+
+module.exports = function(eleventyConfig) {
+    eleventyConfig.addDataExtension("yaml", (contents, filePath) => {
+        console.log("Trying to process ", filePath);
+        return yaml.load(contents);
+    });
```
Comment (tskarhed) [inline .eleventy.js:7]: This doesn't do anything

File: generate_yaml_data.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    metadata.midiFiles = midiFiles;
+    console.log(metadata.midiFiles);
+
+    const svgFiles = globSync(`./songs/${shortName}*.svg`).reverse();
+    if(!svgFiles) {
+        console.log(shortName, 'does not have any SVG files');
+    }
+    metadata.svgFiles = svgFiles;
+
+    const mdFile = globSync(`./songs/${shortName}.md`);
```
Comment (tskarhed) [inline generate_yaml_data.js:47]: Read a yaml file here instead and use the md file as the final output.

File: .eleventy.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+
+    //eleventyConfig.setServerPassthroughCopyBehavior("passthrough");
+
+    // Copy any .svg or midi file to `_site`, via Glob pattern
+    // Keeps the same directory structure
+    eleventyConfig.addPassthroughCopy({"songs/*.svg": "songs"});
+    eleventyConfig.addPassthroughCopy({"songs/*.midi": "songs"});
```
Comment (tskarhed) [inline .eleventy.js:15]: All svgs/midis are stored in the same folder. It is all based on naming convention. This should be changed in the future

File: songs/_includes/main.njk [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    <meta charset="UTF-8">
+    <meta name="viewport"
+          content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
+    <meta http-equiv="X-UA-Compatible" content="ie=edge">
+    <title>scores-ly</title>
+    <link rel="stylesheet" href="/styles.css">
+
+</head>
+<body>
+<a href="/">scoresly</a>
```
Comment (OskarApoex) [inline songs/_includes/main.njk:13]:
```suggestion
<a href="/">Alla låtar</a>
```

File: .eleventy.js (hunk 4 above)
Comment (OskarApoex) [reply .eleventy.js:7]: good

File: .eleventy.js (hunk 6 above)
Comment (OskarApoex) [reply .eleventy.js:15]: yes

**Review by OskarApoex — APPROVED** [review]
Left a small suggestion.
Otherwise it's all good.
Note! The lilypond version must be the same or greater than the highest version of the lilypond files.

File: songs/_includes/main.njk [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    <meta http-equiv="X-UA-Compatible" content="ie=edge">
+    <title>scores-ly</title>
+    <link rel="stylesheet" href="/styles.css">
+
+</head>
+<body>
+<a href="/">Alla låtar</a>
+{{ content | safe}}
+<footer>
+Av Joakim Sandström, Tobias Skarhed och Oskar Tengberg
```
Comment (OskarApoex) [inline songs/_includes/main.njk:16]: image
Kan stylea så den är längst ner på skärmen?

Comment (tskarhed) [reply songs/_includes/main.njk:16]: Äsch, tar det sen

**Pull request merged**